File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/getAmazonProductVariants/GetAmazonProductVariantsHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.getAmazonProductVariants

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.getAmazonProductVariants.GetAmazonProductVariantsApi
import features.getAmazonProductVariants.models.RewardResponseModel

import features.getAmazonProductVariants.serializeGetAmazonProductVariantsState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult


class GetAmazonProductVariantsHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_amazon_product_variants"

    override val eventChannel: String
        get() = "customer_sdk/events/get_amazon_product_variants"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return getAmazonProductVariants(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun getAmazonProductVariants(call: MethodCall): String? {
            val result = GetAmazonProductVariantsApi().getAmazonProductVariants(
                overrideHeaders = call.argument<Map<String, String>>("overrideHeaders") as Map<String, String>?,
                rewardId = call.argument<Int>("rewardId") as Int,
                externalProductId = call.argument<String>("externalProductId") as String
            ).toJsonResult<RewardResponseModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        GetAmazonProductVariantsApi().getAmazonProductVariantsState.serializeGetAmazonProductVariantsState().observe { result ->
            pushSerializedResultToEventSink<GetAmazonProductVariantsApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/getAmazonProductVariants/GetAmazonProductVariantsHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class GetAmazonProductVariantsHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_amazon_product_variants"

    var eventChannel: String = "customer_sdk/events/get_amazon_product_variants"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await GetAmazonProductVariantsHandler.getAmazonProductVariants(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getAmazonProductVariants(call: FlutterMethodCall) async -> String? {


        


        

        guard let args = call.arguments else {
            return nil
        }
        do {
       if let myArgs = args as? [String: Any] {
          let overrideHeaders = myArgs["overrideHeaders"] as? [String: String]
                        guard let rewardId = myArgs["rewardId"] as? Int32 else {return nil}
                guard let externalProductId = myArgs["externalProductId"] as? String else {return nil}

        let result = try await GetAmazonProductVariantsApi().getAmazonProductVariants(
                overrideHeaders: overrideHeaders,
                rewardId: rewardId,
                externalProductId: externalProductId
           ).serializeGetAmazonProductVariantsApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
                }
            }
        }
        catch {
            return nil
        }
        
        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        GetAmazonProductVariantsApi().getAmazonProductVariantsState
            .serializeGetAmazonProductVariantsState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
